src/ServiceControl/Operations/ErrorPersister.cs

```diff
@@ -80,60 +80,52 @@ await SaveToDb(message.Headers.UniqueId(), processingAttempt, groups)
+
+                                var duplicateIndex = _.findIndex(this.{processingAttemptsField}, function(a){{ 
+                                    return a.{attemptedAtField} === attempt.{attemptedAtField};
+                                }});
+
+                                if(duplicateIndex === -1){{
+                                    this.{processingAttemptsField} = _.union(this.{processingAttemptsField}, [attempt]);
+                                }}",
+                    Values = new Dictionary<string, object>
                     {
-                        Name = nameof(FailedMessage.FailureGroups),
-                        Type = PatchCommandType.Set,
-                        Value = RavenJToken.FromObject(groups)
+                        { "status", (int)FailedMessageStatus.Unresolved },
+                        { "failureGroups", RavenJToken.FromObject(groups) },
+                        { "attempt", RavenJToken.FromObject(processingAttempt, Serializer)}
                     }
-                },
-                new[]
+                }, 
+                new ScriptedPatchRequest
                 {
-                    new PatchRequest
-                    {
-                        Name = nameof(FailedMessage.UniqueMessageId),
-                        Type = PatchCommandType.Set,
-                        Value = uniqueMessageId
-                    },
-                    new PatchRequest
-                    {
-                        Name = nameof(FailedMessage.Status),
-                        Type = PatchCommandType.Set,
-                        Value = (int)FailedMessageStatus.Unresolved
-                    },
-                    new PatchRequest
+                    Script = $@"this.{statusField} = status;
+                                this.{failureGroupsField} = failureGroups;
+                                this.{processingAttemptsField} = [attempt];
+                                this.{uniqueMessageIdField} = uniqueMessageId;
+                             ",
+                    Values = new Dictionary<string, object>
                     {
-                        Name = nameof(FailedMessage.ProcessingAttempts),
-                        Type = PatchCommandType.Add,
-                        Value = RavenJToken.FromObject(processingAttempt, Serializer) // Need to specify serilaizer here because otherwise the $type for EndpointDetails is missing and this causes EventDispatcher to blow up!
-                    },
-                    new PatchRequest
-                    {
-                        Name = nameof(FailedMessage.FailureGroups),
-                        Type = PatchCommandType.Set,
-                        Value = RavenJToken.FromObject(groups)
+                        { "status", (int)FailedMessageStatus.Unresolved },
+                        { "failureGroups", RavenJToken.FromObject(groups) },
+                        { "attempt", RavenJToken.FromObject(processingAttempt, Serializer)},
```

Comment:
Is it worth pulling the `failureGroups` and `attempt` definitions out so that we aren't calling `RavenJToken.FromObject` multiple times?